src/test/ui/pattern/bindings-after-at/borrowck-pat-ref-mut-and-ref.rs

```diff
@@ -0,0 +1,87 @@
+#![feature(bindings_after_at)]
+//~^ WARN the feature `bindings_after_at` is incomplete and may cause the compiler to crash
+
+enum Option<T> {
+    None,
+    Some(T),
+}
+
+fn main() {
+    match &mut Some(1) {
+        ref mut z @ &mut Some(ref a) => {
+        //~^ ERROR cannot borrow `_` as immutable because it is also borrowed as mutable
+            **z = None;
+            println!("{}", *a);
+        }
+        _ => ()
+    }
+
+    struct U;
+
+    let ref a @ ref mut b = U; // FIXME: This should not compile.
+    let ref mut a @ ref b = U; // FIXME: This should not compile.
+    let ref a @ (ref mut b, ref mut c) = (U, U); // FIXME: This should not compile.
+    let ref mut a @ (ref b, ref c) = (U, U); // FIXME: This should not compile.
+
+    // FIXME: Seems like we have a soundness hole here.
```

Comment:
N.B. this test.

Reply:
Looks like promotion weirdness.

Reply:
Ah; so this is in fact sound then? (Should not be allowed regardless.)

Reply:
Correct, it's sound, but has very surprising semantics.

Reply:
I wouldn't mind getting a little more clarity here. Is the idea that the shared borrow does not actually alias the mutable borrow, because promotion is duplicating the state?

If for some reason we decide to land this PR without fixing the semantic issue in the compiler (since this is after all feature gated), it would be better if the test demonstrated this concretely, e.g. by adding an `i32` payload (e.g. `struct U(i32)`) and showing that the value of `b` does not get modified by the mutation through `a`.

Reply:
> I wouldn't mind getting a little more clarity here. Is the idea that the shared borrow does not actually alias the mutable borrow, because promotion is duplicating the state?

That's my understanding (but @matthewjasper could confirm) that `let ref mut x @ ref y = U;` will promote `U` twice and so they are disjoint places.

> If for some reason we decide to land this PR without fixing the semantic issue in the compiler (since this is after all feature gated),

I've added the necessary analysis (`check_match::check_borrow_conflicts_in_at_patterns`) to prevent these cases to happen and adjusted the tests accordingly.